Rebase and fix PR #20039

atosatto opened the pull request:
##### SUMMARY
This PR aims to rebase and fix PR #20039 addressing issue #19972.

##### ISSUE TYPE
Bugfix Pull Request

##### COMPONENT NAME
ec2_eni module

##### ANSIBLE VERSION
```
ansible --version                                                                                                                                                                                                                                                     !8598
ansible 2.2.1.0
  config file =
  configured module search path = Default w/o overrides
```


##### ADDITIONAL INFORMATION
This code has been tested with the following playbook in order to simulate different usage scenarios of the `ec2_eni` module.

```
- name: Debug the ec2_eni module
  hosts: all
  vars:
    subnet_id: "<subnet-id>"
  tasks:

    - name: Create a first ENI
      ec2_eni:
        subnet_id: "{{ subnet_id }}"
        private_ip_address: "10.0.1.10"
        description: "First ENI"
        security_groups:
          - a-group
        state: "present"
      register: first_eni

    - name: Create a second ENI
      ec2_eni:
        subnet_id: "{{ subnet_id }}"
        description: "Second ENI"
        security_groups:
          - a-group
        state: "present"

    - name: Modify the first ENI
      ec2_eni:
        eni_id: "{{ first_eni['interface']['id'] }}"
        subnet_id: "{{ subnet_id }}"
        delete_on_termination: yes
        source_dest_check: no

    - name: Create a new ENI with the same IP of the firstENI
      ec2_eni:
        subnet_id: "{{ subnet_id }}"
        private_ip_address: "10.0.1.10"
        security_groups:
          - a-group
```

vlfig committed:
fix eni lookup logic to update them only when univocally specified (Fixes #19972)

remove redundant eni lookup

atosatto committed:
Rename the univocally_find_eni method to uniquely_find_eni

atosatto committed:
Idempotency logic minor fixes

atosatto commented:
cc @xyv and @willthames since they were involved with the original PR.

atosatto committed:
Fixing pep8 reported issues

willthames commented:
I'm not sure I want to look at `ec2_eni` ever again - it's fraught with weird edge cases. Thanks for fixing the obvious problems with #20039 though.

See also #22327 - I'm not sure how much overlap there is.

I'll hope that @wimnat can look at this properly.

atosatto commented:
Ehi @willthames, I think you linked me the wrong PR.. ;) Are you referring to #22377?

willthames commented:
@atosatto yes, well done, you passed my typo-detection test ;)

wimnat commented:
haha yeh this works well until someone doesn't want to use description to identify an ENI. 

The circular hell of people wanting different things from the eni module continues.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -475,26 +473,30 @@ def detach_eni(eni, module):
 
     try:
         filters = {}
-        if subnet_id:
```
You may be right about removing subnet_id from the filters but I'm not sure how this fixes #19972 other than superficially, since we haven't added a filter for description. Can you explain why subnet_id should be removed?

atosatto replied:
Basically the idea is that we might want to have multiple nics on the same subnet and maybe create them in an iteration loop (e.g. to attach multiple VMs to a subnet after the instances creation). With the current implementation it is impossible to do.
Removing the `subnet_id` from the filters allow us to fix #19972 requiring explicitely the `ip_address` or the id to uniquely identify the ENI if we want to update it.

s-hertel replied:
Ahh I understand.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -475,26 +473,30 @@ def detach_eni(eni, module):
             filters['private-ip-address'] = private_ip_address
         else:
-            if instance_id:
+            if instance_id and device_index:
```
Am I wrong in thinking these are always required together? They are used together in create_eni(). Can we add them in a required_together [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ec2_eni.py#L549)?

atosatto replied:
Yep. You're definitely right here. Will push a new commit with the required_together constraint. :)

atosatto replied:
Fixed with 5804772.

atosatto replied:
I remove the `required_together` constraint in bbe47b5 because otherwise we can't support detaching via

```
# Detach an ENI from an instance
- ec2_eni:
    eni_id: eni-xxxxxxx
    instance_id: None
    state: present
```

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -475,26 +473,30 @@ def detach_eni(eni, module):
-            if device_index:
                 filters['attachment.device-index'] = device_index
 
+        if eni_id is None and len(filters) == 0:
```
I'm not sure if you need this if statement. In this case, won't len(eni_result) be 0 and be returned as None on line 502 (now with your changes line [504](https://github.com/ansible/ansible/pull/23404/files#diff-04a83b3209e468ea2a4c3fcd4b4fcb73R504))?

s-hertel replied:
Oh, nevermind, you're right about this. I guess the filters are entirely optional and it will just get all of the enis in this case. It could accidentally work (if you have only a single network interface) if you don't add this.

atosatto replied:
Yes. Exactly! 😄

s-hertel reviewed (changes requested):
Shipit from me after comments are addressed. :)


To people concerned with overlap: I don't think this isn't contrary to any changes in my PR that you linked as we are solving different issues. But this is more closely related to my merged PRs #22689 and #22919 which operated under the idea that either a unique match needs to be found, or if there are multiple results found by whatever filters used, then no modifications should be made. But that fix was decided upon because I was afraid of breaking backward compatibility since this module is stable interface. However, I think this is a superior fix.

Regarding my linked PR, you have the same bug in this uniquely_find_eni() that my pr is fixing: if instance_id is a new id to attach, this filter is going to return None. And then the new instance will never be attached to the correct eni. I think that's fine - different bugs being solved, not overlap.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -475,26 +473,30 @@ def detach_eni(eni, module):
-            filters['subnet-id'] = subnet_id
+
+        # proceed only if we're univocally specifying an ENI
+        if eni_id is None and private_ip_address is None and (instance_id is None and device_index is None):
```
Since a lot of people have been relying on broken functionality (filtering without specifying a unique identifier), perhaps we want to add something to the docs, clarifying how modification should work. The examples for modify do use eni_id though, so perhaps we don't want to add to the docs and instead should just let people know when relevant issues are run across that the functionality only ever would have worked due to chance. Not requesting a change for this, just wondering about your thoughts.

atosatto replied:
I agree with you. I think we should add an example in which, for example, we show how an ENI can be created and edited getting it's id from the ec2_eni facts.
If you agree I will add it to the PR.

s-hertel replied:
I think that would be helpful.

atosatto replied:
I've added an additional example usage example of matching via `subnet_id` and `private_ip_address` in b2a50a6

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -475,26 +473,30 @@ def detach_eni(eni, module):
 
     try:
         filters = {}
-        if subnet_id:
```
Oh, but I guess I'm not sure about this for another reason. private_ip_address isn't unique. Isn't more filtering better in a case like that?

atosatto replied:
You're right. I'm adding a commit to address also that case. Well spotted! :)

atosatto replied:
Fixed in 28ba4c11a and tested with the following playbook snippet.

```
- name: Debug the ec2_eni module
  hosts: all
  vars:
    # NB: The two subnets have the same IP ranges (they can since they're in different VPCs)
    first_subnet_id: "<subnet-id-1>"
    second_subnet_id: "<subnet-id-2>"
  tasks:

    - name: Create a first ENI
      ec2_eni:
        subnet_id: "{{ first_subnet_id }}"
        private_ip_address: "10.0.1.10"
        description: "First ENI"
        state: "present"
      register: first_eni

    - name: Create a second ENI
      ec2_eni:
        subnet_id: "{{ second_subnet_id }}"
        description: "Second ENI"
        private_ip_address: "10.0.1.10"
        state: "present"

    - name: Modify the first ENI via eni_id
      ec2_eni:
        eni_id: "{{ first_eni['interface']['id'] }}"
        subnet_id: "{{ first_subnet_id }}"
        delete_on_termination: yes
        source_dest_check: no

    - name: Modify the second ENI via subnet_id and private_ip_address
      ec2_eni:
        subnet_id: "{{ second_subnet_id }}"
        private_ip_address: "10.0.1.10"
        description: "I should be Second ENI"
```

atosatto committed:
Adding a required_together constraint for instance_id and device_index

atosatto committed:
Fix ec2_eni private_ip_address matching

atosatto committed:
Adding an example of matching via subnet_id and private_ip_address

atosatto committed:
Removing the required_together and subnet_id constraints

atosatto commented:
Ehi @s-hertel I think I've addressed all your feedback.
Could you please do a final review of my PR and tell me if there is anything more you would change/improve before considering it good to ship?
Thank you!

s-hertel commented:
shipit

s-hertel commented:
I don't understand why you removed the required_together for 'instance_id' and 'device_index' but not a blocker. LGTM 💯

s-hertel commented:
@wimnat Knowing this is not the most enjoyable thing to review, I apologetically ask if you want to give some insight on this change. I feel like requiring private_ip_address or eni_id is a step in the right direction since it eliminates some luck of what response you get but you're the expert on this.

atosatto commented:
Ehi @wimnat, can you please help me reviewing my changes to the `ec2_eni` module?! If needed I can provide more test cases. :)

Thanks!

wimnat commented:
So are we saying here that the values we will use to identify an ENI are:

an eni id
or
a private ip address AND subnet id
or
instance id AND device id
?

I'm happy with this.  Can it be documented in the notes section?

Let's get this merged and then i'm going to rewrite as boto3 any way

atosatto commented:
I will! :)

atosatto committed:
Addding some notes to explain the module assumptions

atosatto commented:
@wimnat  please let me know if you have better words to be put in the notes.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -108,6 +108,9 @@
     - aws
     - ec2
+notes:
+    - This modules leverage the uniqueness of end_id, private_ip_address and subnet_id,
```
end_id -> eni_id

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -108,6 +108,9 @@
     - aws
     - ec2
+notes:
+    - This modules leverage the uniqueness of end_id, private_ip_address and subnet_id,
```
How about "This module identifies and ENI based on either the eni_id, a combination of private_ip_address and subnet_id, or a combination of instance_id and device_id. Any of these options will let you specify a particular ENI." ?

atosatto replied:
+1

atosatto committed:
Implementing @ryansb's wording proposal

ryansb commented:
Thank you @atosatto, and congrats on your first contribution!

zyv commented:
@atosatto ok I was not casted correctly, so I didn't get any notifications and thus didn't test whether it still works correctly for me, but I hope it does, and if not I know whom I have to complain to :)
